quinn-proto/src/config/mod.rs

```diff
@@ -503,3 +518,22 @@ impl From<VarIntBoundsExceeded> for ConfigError {
         Self::OutOfBounds
     }
 }
+
+/// Object to get current [`SystemTime`]
+///
+/// This exists to allow system time to be mocked in tests, or wherever else desired.
+pub trait SystemTimeClock: Send + Sync {
```

Comment:
> I don't love these names. How about TimeSource for the trait and StdSystemTime for the default implementation?

Sure. Done.

> Have you actually tried to mock this?

No.

> IMO that could be unergonomic because SystemTime doesn't have any public constructors, so we're limited to checked_sub() and checked_add() calls from either now() or UNIX_EPOCH.

IMO, constructing `SystemTime` for a given unix time via `UNIX_EPOCH + Duration::from_secs(unix_secs)` or what have you is blase enough. `Token::from_bytes` does it plus a byte read and error handling in a single expression, and doesn't even factor it out. Yet.

> Maybe we should use a custom type for RetryToken::issued and also build the API to be more specific around token issuance and expiration checking?

I don't love the idea of trying to highly abstract out only the mathematically relevant aspects of it. For one, we may want to pass these `SystemTime` in to a `TokenLog`, in which case it could be hard to predict what exact methods on `SystemTime` it would find useful. For example, that's critical functionality to `BloomTokenLog`.

Eventually, I would like to work towards the ability to sort of just more manually send and process tokens at the application level, in various ways that would require various different PRs. Moving it in that direction could sort of make a desire to have a "custom expiration logic object" obsolete as the things it lets you do would enclose it.

One thing that's a bit related to this is the fact that, as of #1996, one must import `crate::{Duration, Instant, SystemTime, UNIX_EPOCH}` rather than `std::time::{Duration, Instant, SystemTime, UNIX_EPOCH}`, or the code will initially appear to work but then eventually get a CI error when compiling for WASM. That's a bit of an annoyance. In my `time-test-1` branch I [experimented](https://github.com/gretchenfrage/quinn/commit/1e64f249d38192f36d75b024a1987b3a769a470a) with trying to wrap `Duration`/`Instant`/`SystemTime` within newtypes that conditionally compile whether they're wrappers around `std::time::X` or `web_time::X`, so this would be less of a problem. However, it was sort of a huge pain to be manually writing wrapper boilerplate for every single method and `std::op` trait impl on `Duration`/`Instant`/`SystemTime` we use.

Tangentially related: As of #1996, proto conditionally compiles whether whether it uses `std::time::X` types or `web_time::X` types in `lib.rs`:

```rs
// Deal with time
#[cfg(not(all(target_family = "wasm", target_os = "unknown")))]
pub(crate) use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
#[cfg(all(target_family = "wasm", target_os = "unknown"))]
pub(crate) use web_time::{Duration, Instant, SystemTime, UNIX_EPOCH};
```

But this is not a completely internal thing, as there are various places in proto's public API which expose these types. For example, every place in proto where a method accepts a `now: Instant`. In terms of features being strictly additive to the API, that seems bad?

Reply:
> In terms of features being strictly additive to the API, that seems bad?

Mutually exclusive `cfg`s aren't features, and needn't be additive. We should still try to minimize surprise, of course.